Crash during pure SSH object transfer with multiple objects

**Describe the bug**
When cloning a repository with multiple Git LFS objects using the "pure" SSH version of the Git LFS transfer [protocol](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/docs/proposals/ssh_adapter.md), the client may crash due to a `nil` pointer reference.

**To Reproduce**
To reproduce the behaviour we can use a modified version of the `batch transfers with ssh endpoint (git-lfs-transfer)` [test](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/t/t-batch-transfer.sh#L166-L192) in the `t/t-batch-transfer.sh` test script.

```shell
begin_test "batch transfers with ssh endpoint and multiple objects (git-lfs-transfer)"
(
  set -e

  setup_pure_ssh

  reponame="batch-ssh-transfer-multiple"
  setup_remote_repo "$reponame"
  clone_repo "$reponame" "$reponame"

  git lfs track "*.dat"
  printf "test1" >test1.dat
  printf "test2" >test2.dat
  git add .gitattributes test*.dat
  git commit -m "initial commit"

  sshurl=$(ssh_remote "$reponame")
  git config "lfs.url" "$sshurl"

  # this is required to avoid a bug with our lfs-ssh-echo.go utility
  git config "lfs.ssh.automultiplex" "false"

  git push origin main

  cd ..

  git clone "$sshurl" "$reponame-2"
)
end_test
```

Running this test script results in the following crash and stack trace:
```
Cloning into 'batch-ssh-transfer-multiple-2'...
Downloading test2.dat (5 B) 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x10307df04]

goroutine 86 [running]:
github.com/git-lfs/git-lfs/v3/ssh.(*PktlineConnection).Lock(...)
        github.com/git-lfs/git-lfs/v3/ssh/protocol.go:23
github.com/git-lfs/git-lfs/v3/tq.(*SSHBatchClient).batchInternal(0x14000569a30?, {0x14000510080, 0x3, 0x4}, {0x14000500240, 0x1, 0x1}) 
        github.com/git-lfs/git-lfs/v3/tq/ssh.go:32 +0x64
github.com/git-lfs/git-lfs/v3/tq.(*SSHBatchClient).Batch(0x140004226f0, {0x103895878?, 0x60?}, 0x140005481e0)
        github.com/git-lfs/git-lfs/v3/tq/ssh.go:66 +0x3bc
github.com/git-lfs/git-lfs/v3/tq.Batch({0x10339d9b8, 0x140000c1bc0}, 0x1, {0x14000206b60, 0x6}, 0x14000518300, {0x14000560008, 0x1, 0x1}) 
        github.com/git-lfs/git-lfs/v3/tq/api.go:45 +0x228
github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).enqueueAndCollectRetriesFor(0x1400052e100, {0x14000532708, 0x1, 0x0?})
        github.com/git-lfs/git-lfs/v3/tq/transfer_queue.go:581 +0x25c
github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).collectBatches.func1()
        github.com/git-lfs/git-lfs/v3/tq/transfer_queue.go:470 +0x74
created by github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).collectBatches in goroutine 85
        github.com/git-lfs/git-lfs/v3/tq/transfer_queue.go:463 +0x228
error: external filter 'git-lfs filter-process' failed
fatal: test2.dat: smudge filter lfs failed
```

**Expected behaviour**
All Git LFS objects should be cloned successfully.

**System environment**
The examples above were performed on macOS, but should behave the same way on any platform.

**Output of `git lfs env`**
```
git-lfs/3.5.0 (GitHub; darwin arm64; go 1.23.1; git 8b2882b7)
git version 2.46.0
```

Note that this is a custom build from the current `HEAD` of our `main` development branch, for the `3.5.x` release series.

**Additional context**
This crash may be one which prompted the development of PR #5063.

This problem may also be the source of the issue described in https://github.com/git-lfs/git-lfs/pull/5816#issuecomment-2207214023.

---

The issue appears to stem from the fact that we [shut down](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/tq/transfer_queue.go#L974) our SSH connections when the `Wait()` method of our `TransferQueue` structure runs, and [set](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/tq/transfer_queue.go#L975) the `SSHTransfer` field of the associated `Manifest` structure to `nil`, but a pointer to the same structure has been [saved](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/tq/manifest.go#L250) in the `SSHBatchClient` structure in the manifest, and so when the manifest is [reissued](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/commands/commands.go#L64-L69) by the `getTransferManifestOperationRemote()` function, there is still a non-`nil` reference to the `SSHTransfer` structure that now has a `nil` array of connections, and that is what is [used](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/tq/api.go#L45-L51) for the next batch transfer operation.

Specifically, we [call](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/commands/command_filter_process.go#L106-L111) `NewTransferQueue()` once when smudging in our `git lfs filter-process` command, and [pass](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/commands/command_filter_process.go#L112) the transfer queue to our `infiniteTransferBuffer()` function.  To construct the `manifest` argument of `NewTransferQueue()`, we [call](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/commands/command_filter_process.go#L108) `getTransferManifestOperationRemote()`, which constructs the `Manifest` for the first time.  

We subsequently also [call](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/commands/command_filter_process.go#L131) the `smudge()` function, which also [calls](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/commands/command_smudge.go#L128) `getTransferManifestOperationRemote()`.  That second call will return the existing `Manifest` object created by the previous call; however, at this point the `list_available_blobs` filter [command](https://github.com/git/git/blob/90fe3800b92a49173530828c0a17951abd30f0e1/Documentation/gitattributes.txt#L627-L648) has been received from Git, so we have [called](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/commands/command_filter_process.go#L153) `Wait()` on the transfer queue.  That method, because the manifest has been upgraded to a concrete one with a non-`nil` `SSHTransfer` field, [calls](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/tq/transfer_queue.go#L974) `Shutdown()` on that `SSHTransfer`, which [sets](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/ssh/connection.go#L200) the SSH connection count to zero, which results in the `conn` array being [set](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/ssh/connection.go#L192) to `nil`.

When the manifest is reused to process the batch of objects to be downloaded, `Connection(0)` is [called](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/tq/ssh.go#L28) on the `SSHTransfer` structure pointed to by the `SSHBatchClient` structure, which [returns](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/ssh/connection.go#L103) `nil, nil` because zero is equal to `len(tr.conn)` when `tr.conn` is `nil`.  Since a `nil` is returned for the error value, no error is [reported](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/tq/ssh.go#L29-L31) and instead the code tries to [call](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/tq/ssh.go#L32) `Lock()` on the `nil` pointer, which causes a panic condition.

---

One immediate corrective will be to return a non-`nil` error when `tr.conn` is `nil`.  (A wrinkle here is that we'll also need to rename the `tr` receiver variable to something else so it doesn't override our imported `tr` package name.)  That alone, however, will still result in a failed clone, just not due to a program crash.

Ultimately, we may want to avoid trying to shut down our SSH connections at the end of individual batch operations, and do some analysis of the preferred lifetime for them.  (Among other issues, when pushing, we create two separate `ControlMaster=yes` connections, one for locking operations and another for batch operations; it would be nice to avoid this duplication.)

Shutting down SSH connections cleanly at the end of the entire Git LFS process may also entail some refactoring of how we cause the program to exit.  Right now, we make a token effort at [closing](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/commands/run.go#L155) resources associated with our API clients, but really that just amounts to [closing](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/lfshttp/client.go#L205) our logger, and when we call `os.Exit()` in several dozen places, that step is skipped and we let Go take care of closing open I/O handles.  If we want to always gracefully close multiple open SSH connections by [sending](https://github.com/git-lfs/git-lfs/blob/8b2882b7b791635d5deefebcc4edaa9e8eef638a/ssh/protocol.go#L39) a `quit` message, though, we may have to refactor these `os.Exit()` calls into a common exit routine.

I'm actually getting this crash on my existing repositories after moving them to the pure SSH LFS protocol.
If someone has a workaround that would be great.

```
Cloning into 'XXXXX'...
remote: Enumerating objects: 6629, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 6629 (delta 2), reused 0 (delta 0), pack-reused 6623 (from 1)
Receiving objects: 100% (6629/6629), 7.49 MiB | 4.41 MiB/s, done.
Resolving deltas: 100% (4792/4792), done.
Downloading YYYYY (1.7 MB)
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x1 addr=0x20 pc=0xe9bfd9]

goroutine 68 [running]:
github.com/git-lfs/git-lfs/v3/ssh.(*PktlineConnection).Lock(...)
        github.com/git-lfs/git-lfs/v3/ssh/protocol.go:23
github.com/git-lfs/git-lfs/v3/tq.(*SSHBatchClient).batchInternal(0xc000147a40?, {0xc00019a8c0, 0x3, 0x4}, {0xc000032b40, 0x1, 0x1})
        github.com/git-lfs/git-lfs/v3/tq/ssh.go:32 +0xb9
github.com/git-lfs/git-lfs/v3/tq.(*SSHBatchClient).Batch(0xc00002def0?, {0x60?, 0x60?}, 0xc00005bd40)
        github.com/git-lfs/git-lfs/v3/tq/ssh.go:66 +0x4d9
github.com/git-lfs/git-lfs/v3/tq.Batch({0x11822b8, 0xc00019ab00}, 0x1, {0xc0001f47a0, 0x6}, 0x1?, {0xc00004e090, 0x1, 0x1})
        github.com/git-lfs/git-lfs/v3/tq/api.go:45 +0x2b3
github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).enqueueAndCollectRetriesFor(0xc0003a2200, {0xc000333c00, 0x1, 0x0?})
        github.com/git-lfs/git-lfs/v3/tq/transfer_queue.go:581 +0x2fe
github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).collectBatches.func1()
        github.com/git-lfs/git-lfs/v3/tq/transfer_queue.go:470 +0x85
created by github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).collectBatches in goroutine 67
        github.com/git-lfs/git-lfs/v3/tq/transfer_queue.go:463 +0x29b
error: external filter 'git-lfs filter-process' failed
fatal: YYYYY: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'

```


```
>git lfs env
git-lfs/3.5.1 (GitHub; windows amd64; go 1.21.7; git e237bb3a)
git version 2.47.0.windows.1

LocalWorkingDir=
LocalGitDir=
LocalGitStorageDir=
LocalMediaDir=lfs\objects
LocalReferenceDirs=
TempDir=lfs\tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneVerifyUnreachableAlways=false
PruneRemoteName=origin
LfsStorageDir=lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
GIT_TRACE=0
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

--- a/docs/man/git-lfs-config.adoc
+++ b/docs/man/git-lfs-config.adoc
@@ -91,7 +91,8 @@ keepalive connections. Default: 30 minutes.
 +
 When using the pure SSH-based protocol, whether to multiplex requests
 over a single connection when possible. This option requires the use of
-OpenSSH or a compatible SSH client. Default: true.
+OpenSSH or a compatible SSH client. Default: false on Windows, otherwise
+true.
 * `lfs.ssh.retries`
 +
 Specifies the number of times Git LFS will attempt to obtain

--- a/lfsapi/lfsapi.go
+++ b/lfsapi/lfsapi.go
@@ -62,14 +62,14 @@ func (c *Client) SSHTransfer(operation, remote string) *ssh.SSHTransfer {
 	}
 	uc := config.NewURLConfig(c.context.GitEnv())
 	if val, ok := uc.Get("lfs", endpoint.OriginalUrl, "sshtransfer"); ok && val != "negotiate" && val != "always" {
-		tracerx.Printf("skipping pure SSH protocol connection by request")
+		tracerx.Printf("skipping pure SSH protocol connection by request (%s, %s)", operation, remote)
 		return nil
 	}
 	ctx := c.Context()
-	tracerx.Printf("attempting pure SSH protocol connection")
+	tracerx.Printf("attempting pure SSH protocol connection (%s, %s)", operation, remote)
 	sshTransfer, err := ssh.NewSSHTransfer(ctx.OSEnv(), ctx.GitEnv(), &endpoint.SSHMetadata, operation)
 	if err != nil {
-		tracerx.Printf("pure SSH protocol connection failed: %s", err)
+		tracerx.Printf("pure SSH protocol connection failed (%s, %s): %s", operation, remote, err)
 		return nil
 	}
 	return sshTransfer

--- a/ssh/connection.go
+++ b/ssh/connection.go
@@ -90,67 +90,67 @@ func startConnection(id int, osEnv config.Environment, gitEnv config.Environment
 
 // Connection returns the nth connection (starting from 0) in this transfer
 // instance or nil if there is no such item.
-func (tr *SSHTransfer) IsMultiplexingEnabled() bool {
-	return tr.multiplexing
+func (st *SSHTransfer) IsMultiplexingEnabled() bool {
+	return st.multiplexing
 }
 
 // Connection returns the nth connection (starting from 0) in this transfer
 // instance if it is initialized and otherwise initializes a new connection and
 // saves it in the nth position.  In all cases, nil is returned if n is greater
 // than the maximum number of connections.
-func (tr *SSHTransfer) Connection(n int) (*PktlineConnection, error) {
-	tr.lock.RLock()
-	if n >= len(tr.conn) {
-		tr.lock.RUnlock()
+func (st *SSHTransfer) Connection(n int) (*PktlineConnection, error) {
+	st.lock.RLock()
+	if n >= len(st.conn) {
+		st.lock.RUnlock()
 		return nil, nil
 	}
-	if tr.conn[n] != nil {
-		defer tr.lock.RUnlock()
-		return tr.conn[n], nil
+	if st.conn[n] != nil {
+		defer st.lock.RUnlock()
+		return st.conn[n], nil
 	}
-	tr.lock.RUnlock()
+	st.lock.RUnlock()
 
-	tr.lock.Lock()
-	defer tr.lock.Unlock()
-	if tr.conn[n] != nil {
-		return tr.conn[n], nil
+	st.lock.Lock()
+	defer st.lock.Unlock()
+	if st.conn[n] != nil {
+		return st.conn[n], nil
 	}
-	conn, _, err := tr.spawnConnection(n)
+	conn, _, err := st.spawnConnection(n)
 	if err != nil {
 		return nil, err
 	}
-	tr.conn[n] = conn
+	st.conn[n] = conn
 	return conn, nil
 }
 
 // ConnectionCount returns the number of connections this object has.
-func (tr *SSHTransfer) ConnectionCount() int {
-	tr.lock.RLock()
-	defer tr.lock.RUnlock()
-	return len(tr.conn)
+func (st *SSHTransfer) ConnectionCount() int {
+	st.lock.RLock()
+	defer st.lock.RUnlock()
+	return len(st.conn)
 }
 
 // SetConnectionCount sets the number of connections to the specified number.
-func (tr *SSHTransfer) SetConnectionCount(n int) error {
-	tr.lock.Lock()
-	defer tr.lock.Unlock()
-	return tr.setConnectionCount(n)
+func (st *SSHTransfer) SetConnectionCount(n int) error {
+	st.lock.Lock()
+	defer st.lock.Unlock()
+	return st.setConnectionCount(n)
 }
 
 // SetConnectionCountAtLeast sets the number of connections to be not less than
 // the specified number.
-func (tr *SSHTransfer) SetConnectionCountAtLeast(n int) error {
-	tr.lock.Lock()
-	defer tr.lock.Unlock()
-	count := len(tr.conn)
+func (st *SSHTransfer) SetConnectionCountAtLeast(n int) error {
+	st.lock.Lock()
+	defer st.lock.Unlock()
+	count := len(st.conn)
 	if n <= count {
 		return nil
 	}
-	return tr.setConnectionCount(n)
+	return st.setConnectionCount(n)
 }
 
-func (tr *SSHTransfer) spawnConnection(n int) (*PktlineConnection, string, error) {
-	conn, _, controlPath, err := startConnection(n, tr.osEnv, tr.gitEnv, tr.meta, tr.operation, tr.controlPath)
+func (st *SSHTransfer) spawnConnection(n int) (*PktlineConnection, string, error) {
+	conn, _, controlPath, err := startConnection(n, st.osEnv, st.gitEnv, st.meta, st.operation, st.controlPath)
 	if err != nil {
 		tracerx.Printf("failed to spawn pure SSH connection (#%d): %s", n, err)
 		return nil, "", err
@@ -158,14 +158,14 @@ func (tr *SSHTransfer) spawnConnection(n int) (*PktlineConnection, string, error
 	return conn, controlPath, err
 }
 
-func (tr *SSHTransfer) setConnectionCount(n int) error {
-	count := len(tr.conn)
+func (st *SSHTransfer) setConnectionCount(n int) error {
+	count := len(st.conn)
 	if n < count {
 		tn := n
 		if tn == 0 {
 			tn = 1
 		}
-		for i, item := range tr.conn[tn:count] {
+		for i, item := range st.conn[tn:count] {
 			if item == nil {
 				tracerx.Printf("skipping uninitialized lazy pure SSH connection (#%d) (resetting total from %d to %d)", i, count, n)
 				continue
@@ -175,33 +175,33 @@ func (tr *SSHTransfer) setConnectionCount(n int) error {
 				return err
 			}
 		}
-		tr.conn = tr.conn[0:tn]
+		st.conn = st.conn[0:tn]
 	} else if n > count {
 		for i := count; i < n; i++ {
 			if i == 0 {
-				conn, controlPath, err := tr.spawnConnection(i)
+				conn, controlPath, err := st.spawnConnection(i)
 				if err != nil {
 					return err
 				}
-				tr.conn = append(tr.conn, conn)
-				tr.controlPath = controlPath
+				st.conn = append(st.conn, conn)
+				st.controlPath = controlPath
 			} else {
-				tr.conn = append(tr.conn, nil)
+				st.conn = append(st.conn, nil)
 			}
 		}
 	}
 	if n == 0 && count > 0 {
 		tracerx.Printf("terminating pure SSH connection (#0) (resetting total from %d to %d)", count, n)
-		if err := tr.conn[0].End(); err != nil {
+		if err := st.conn[0].End(); err != nil {
 			return err
 		}
-		tr.conn = nil
-		tr.controlPath = ""
+		st.conn = nil
+		st.controlPath = ""
 	}
 	return nil
 }
 
-func (tr *SSHTransfer) Shutdown() error {
+func (st *SSHTransfer) Shutdown() error {
 	tracerx.Printf("shutting down pure SSH connections")
-	return tr.SetConnectionCount(0)
+	return st.SetConnectionCount(0)
 }

--- a/ssh/connection.go
+++ b/ssh/connection.go
@@ -96,13 +96,14 @@ func (st *SSHTransfer) IsMultiplexingEnabled() bool {
 
 // Connection returns the nth connection (starting from 0) in this transfer
 // instance if it is initialized and otherwise initializes a new connection and
-// saves it in the nth position.  In all cases, nil is returned if n is greater
-// than the maximum number of connections.
+// saves it in the nth position.  In all cases, nil is returned with an error
+// if n is greater than the maximum number of connections, including when
+// the connection array itself is nil.
 func (st *SSHTransfer) Connection(n int) (*PktlineConnection, error) {
 	st.lock.RLock()
 	if n >= len(st.conn) {
 		st.lock.RUnlock()
-		return nil, nil
+		return nil, errors.New(tr.Tr.Get("pure SSH connection unavailable (#%d)", n))
 	}
 	if st.conn[n] != nil {
 		defer st.lock.RUnlock()

--- a/ssh/connection.go
+++ b/ssh/connection.go
@@ -42,7 +42,7 @@ func NewSSHTransfer(osEnv config.Environment, gitEnv config.Environment, meta *S
 }
 
 func startConnection(id int, osEnv config.Environment, gitEnv config.Environment, meta *SSHMetadata, operation string, multiplexControlPath string) (conn *PktlineConnection, multiplexing bool, controlPath string, err error) {
-	tracerx.Printf("spawning pure SSH connection")
+	tracerx.Printf("spawning pure SSH connection (#%d)", id)
 	var errbuf bytes.Buffer
 	exe, args, multiplexing, controlPath := GetLFSExeAndArgs(osEnv, gitEnv, meta, "git-lfs-transfer", operation, true, multiplexControlPath)
 	cmd, err := subprocess.ExecCommand(exe, args...)
@@ -81,8 +81,10 @@ func startConnection(id int, osEnv config.Environment, gitEnv config.Environment
 		w.Close()
 		cmd.Wait()
 		err = errors.Combine([]error{err, fmt.Errorf(tr.Tr.Get("Failed to connect to remote SSH server: %s", cmd.Stderr))})
+		tracerx.Printf("pure SSH connection unsuccessful (#%d)", id)
+	} else {
+		tracerx.Printf("pure SSH connection successful (#%d)", id)
 	}
-	tracerx.Printf("pure SSH connection successful")
 	return conn, multiplexing, controlPath, err
 }
 
@@ -150,7 +152,7 @@ func (tr *SSHTransfer) SetConnectionCountAtLeast(n int) error {
 func (tr *SSHTransfer) spawnConnection(n int) (*PktlineConnection, string, error) {
 	conn, _, controlPath, err := startConnection(n, tr.osEnv, tr.gitEnv, tr.meta, tr.operation, tr.controlPath)
 	if err != nil {
-		tracerx.Printf("failed to spawn pure SSH connection: %s", err)
+		tracerx.Printf("failed to spawn pure SSH connection (#%d): %s", n, err)
 		return nil, "", err
 	}
 	return conn, controlPath, err
@@ -163,12 +165,12 @@ func (tr *SSHTransfer) setConnectionCount(n int) error {
 		if tn == 0 {
 			tn = 1
 		}
-		for _, item := range tr.conn[tn:count] {
+		for i, item := range tr.conn[tn:count] {
 			if item == nil {
-				tracerx.Printf("skipping uninitialized lazy pure SSH connection (%d -> %d)", count, n)
+				tracerx.Printf("skipping uninitialized lazy pure SSH connection (#%d) (resetting total from %d to %d)", i, count, n)
 				continue
 			}
-			tracerx.Printf("terminating pure SSH connection (%d -> %d)", count, n)
+			tracerx.Printf("terminating pure SSH connection (#%d) (resetting total from %d to %d)", tn+i, count, n)
 			if err := item.End(); err != nil {
 				return err
 			}
@@ -189,7 +191,7 @@ func (tr *SSHTransfer) setConnectionCount(n int) error {
 		}
 	}
 	if n == 0 && count > 0 {
-		tracerx.Printf("terminating pure SSH connection (%d -> %d)", count, n)
+		tracerx.Printf("terminating pure SSH connection (#0) (resetting total from %d to %d)", count, n)
 		if err := tr.conn[0].End(); err != nil {
 			return err
 		}
@@ -200,6 +202,6 @@ func (tr *SSHTransfer) setConnectionCount(n int) error {
 }
 
 func (tr *SSHTransfer) Shutdown() error {
-	tracerx.Printf("shutting down pure SSH connection")
+	tracerx.Printf("shutting down pure SSH connections")
 	return tr.SetConnectionCount(0)
 }

--- a/ssh/ssh.go
+++ b/ssh/ssh.go
@@ -30,7 +30,7 @@ type SSHMetadata struct {
 	Path        string
 }
 
-func FormatArgs(cmd string, args []string, needShell bool, multiplex bool, controlPath string) (string, []string) {
+func FormatArgs(cmd string, args []string, needShell bool) (string, []string) {
 	if !needShell {
 		return cmd, args
 	}
@@ -41,7 +41,7 @@ func FormatArgs(cmd string, args []string, needShell bool, multiplex bool, contr
 func GetLFSExeAndArgs(osEnv config.Environment, gitEnv config.Environment, meta *SSHMetadata, command, operation string, multiplexDesired bool, multiplexControlPath string) (exe string, args []string, multiplexing bool, controlPath string) {
 	exe, args, needShell, multiplexing, controlPath := GetExeAndArgs(osEnv, gitEnv, meta, multiplexDesired, multiplexControlPath)
 	args = append(args, fmt.Sprintf("%s %s %s", command, meta.Path, operation))
-	exe, args = FormatArgs(exe, args, needShell, multiplexing, controlPath)
+	exe, args = FormatArgs(exe, args, needShell)
 	tracerx.Printf("run_command: %s %s", exe, strings.Join(args, " "))
 	return exe, args, multiplexing, controlPath
 }

--- a/ssh/ssh_test.go
+++ b/ssh/ssh_test.go
@@ -45,7 +45,8 @@ func TestSSHGetExeAndArgsSsh(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "ssh", exe)
 	assert.Equal(t, []string{"user@foo.com"}, args)
 }
@@ -61,7 +62,8 @@ func TestSSHGetExeAndArgsSshCustomPort(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "ssh", exe)
 	assert.Equal(t, []string{"-p", "8888", "user@foo.com"}, args)
 }
@@ -79,7 +81,7 @@ func TestSSHGetExeAndArgsSshNoMultiplexing(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 
 	exe, baseargs, needShell, multiplexing, controlPath := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, true, "")
-	exe, args := ssh.FormatArgs(exe, baseargs, needShell, multiplexing, controlPath)
+	exe, args := ssh.FormatArgs(exe, baseargs, needShell)
 	assert.Equal(t, "ssh", exe)
 	assert.Equal(t, false, multiplexing)
 	assert.Equal(t, []string{"user@foo.com"}, args)
@@ -99,7 +101,7 @@ func TestSSHGetExeAndArgsSshMultiplexingMaster(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 
 	exe, baseargs, needShell, multiplexing, controlPath := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, true, "")
-	exe, args := ssh.FormatArgs(exe, baseargs, needShell, multiplexing, controlPath)
+	exe, args := ssh.FormatArgs(exe, baseargs, needShell)
 	assert.Equal(t, "ssh", exe)
 	assert.Equal(t, true, multiplexing)
 	assert.Equal(t, 3, len(args))
@@ -122,7 +124,7 @@ func TestSSHGetExeAndArgsSshMultiplexingExtra(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 
 	exe, baseargs, needShell, multiplexing, controlPath := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, true, "/tmp/lfs/lfs.sock")
-	exe, args := ssh.FormatArgs(exe, baseargs, needShell, multiplexing, controlPath)
+	exe, args := ssh.FormatArgs(exe, baseargs, needShell)
 	assert.Equal(t, "ssh", exe)
 	assert.Equal(t, true, multiplexing)
 	assert.Equal(t, []string{"-oControlMaster=no", "-oControlPath=/tmp/lfs/lfs.sock", "user@foo.com"}, args)
@@ -141,7 +143,8 @@ func TestSSHGetExeAndArgsPlink(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"user@foo.com"}, args)
 }
@@ -159,7 +162,8 @@ func TestSSHGetExeAndArgsPlinkCustomPort(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-P", "8888", "user@foo.com"}, args)
 }
@@ -178,7 +182,8 @@ func TestSSHGetExeAndArgsPlinkCustomPortExplicitEnvironment(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-P", "8888", "user@foo.com"}, args)
 }
@@ -197,7 +202,8 @@ func TestSSHGetExeAndArgsPlinkCustomPortExplicitEnvironmentPutty(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-P", "8888", "user@foo.com"}, args)
 }
@@ -216,7 +222,8 @@ func TestSSHGetExeAndArgsPlinkCustomPortExplicitEnvironmentSsh(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-p", "8888", "user@foo.com"}, args)
 }
@@ -233,7 +240,8 @@ func TestSSHGetExeAndArgsTortoisePlink(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-batch", "user@foo.com"}, args)
 }
@@ -251,7 +259,8 @@ func TestSSHGetExeAndArgsTortoisePlinkCustomPort(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-batch", "-P", "8888", "user@foo.com"}, args)
 }
@@ -270,7 +279,8 @@ func TestSSHGetExeAndArgsTortoisePlinkCustomPortExplicitEnvironment(t *testing.T
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-batch", "-P", "8888", "user@foo.com"}, args)
 }
@@ -291,7 +301,8 @@ func TestSSHGetExeAndArgsTortoisePlinkCustomPortExplicitConfig(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-batch", "-P", "8888", "user@foo.com"}, args)
 }
@@ -311,7 +322,8 @@ func TestSSHGetExeAndArgsTortoisePlinkCustomPortExplicitConfigOverride(t *testin
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, plink, exe)
 	assert.Equal(t, []string{"-P", "8888", "user@foo.com"}, args)
 }
@@ -327,7 +339,8 @@ func TestSSHGetExeAndArgsSshCommandPrecedence(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", "sshcmd user@foo.com"}, args)
 }
@@ -342,7 +355,8 @@ func TestSSHGetExeAndArgsSshCommandArgs(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", "sshcmd --args 1 user@foo.com"}, args)
 }
@@ -357,7 +371,8 @@ func TestSSHGetExeAndArgsSshCommandArgsWithMixedQuotes(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", "sshcmd foo 'bar \"baz\"' user@foo.com"}, args)
 }
@@ -372,7 +387,8 @@ func TestSSHGetExeAndArgsSshCommandCustomPort(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", "sshcmd -p 8888 user@foo.com"}, args)
 }
@@ -388,7 +404,8 @@ func TestSSHGetExeAndArgsCoreSshCommand(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", "sshcmd --args 2 user@foo.com"}, args)
 }
@@ -402,7 +419,8 @@ func TestSSHGetExeAndArgsCoreSshCommandArgsWithMixedQuotes(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", "sshcmd foo 'bar \"baz\"' user@foo.com"}, args)
 }
@@ -416,7 +434,8 @@ func TestSSHGetExeAndArgsConfigVersusEnv(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", "sshcmd --args 1 user@foo.com"}, args)
 }
@@ -432,7 +451,8 @@ func TestSSHGetExeAndArgsPlinkCommand(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", plink + " user@foo.com"}, args)
 }
@@ -449,7 +469,8 @@ func TestSSHGetExeAndArgsPlinkCommandCustomPort(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", plink + " -P 8888 user@foo.com"}, args)
 }
@@ -465,7 +486,8 @@ func TestSSHGetExeAndArgsTortoisePlinkCommand(t *testing.T) {
 	meta := ssh.SSHMetadata{}
 	meta.UserAndHost = "user@foo.com"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", plink + " -batch user@foo.com"}, args)
 }
@@ -482,7 +504,8 @@ func TestSSHGetExeAndArgsTortoisePlinkCommandCustomPort(t *testing.T) {
 	meta.UserAndHost = "user@foo.com"
 	meta.Port = "8888"
 
-	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, ""))
+	exe, args, needShell, _, _ := ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false, "")
+	exe, args = ssh.FormatArgs(exe, args, needShell)
 	assert.Equal(t, "sh", exe)
 	assert.Equal(t, []string{"-c", plink + " -batch -P 8888 user@foo.com"}, args)
 }

--- a/t/cmd/lfs-ssh-echo.go
+++ b/t/cmd/lfs-ssh-echo.go
@@ -73,8 +73,12 @@ func main() {
 		}
 		if pathArg, found := strings.CutPrefix(os.Args[offset+1], "-oControlPath="); found {
 			if master {
-				if file, err := os.OpenFile(pathArg, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0); err != nil {
-					fmt.Fprintf(os.Stderr, "expected %q to not exist", pathArg)
+				if file, err := os.OpenFile(pathArg, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0666); err != nil {
+					if os.IsExist(err) {
+						fmt.Fprintf(os.Stderr, "expected %q to not exist", pathArg)
+					} else {
+						fmt.Fprintf(os.Stderr, "unable to create %q: %s", pathArg, err)
+					}
 					os.Exit(1)
 				} else {
 					file.Close()
@@ -82,7 +86,11 @@ func main() {
 				}
 			} else {
 				if file, err := os.OpenFile(pathArg, os.O_RDONLY, 0); err != nil {
-					fmt.Fprintf(os.Stderr, "expected %q to exist", pathArg)
+					if os.IsNotExist(err) {
+						fmt.Fprintf(os.Stderr, "expected %q to exist", pathArg)
+					} else {
+						fmt.Fprintf(os.Stderr, "unable to open %q: %s", pathArg, err)
+					}
 					os.Exit(1)
 				} else {
 					file.Close()

--- a/t/t-batch-transfer.sh
+++ b/t/t-batch-transfer.sh
@@ -165,6 +165,53 @@ begin_test "batch transfers with ssh endpoint (git-lfs-authenticate)"
 )
 end_test
 
+assert_ssh_transfer_session_counts() {
+  local log="$1"
+  local msg="$2"
+  local min="$3"
+  local max="$4"
+
+  local count="$(grep -c "$msg" "$log")"
+
+  [ "$max" -ge "$count" ]
+  [ "$min" -le "$count" ]
+}
+
+assert_ssh_transfer_sessions() {
+  local log="$1"
+  local direction="$2"
+  local num_objs="$3"
+  local objs_per_batch="$4"
+
+  local min_expected_start=1
+  local max_expected_start=$(( num_objs > objs_per_batch ? objs_per_batch : num_objs ))
+  local min_expected_end=1
+  local max_expected_end="$max_expected_start"
+
+  local expected_ctrl=1
+
+  # On upload we currently spawn one extra control socket SSH connection
+  # to run locking commands and never shut it down cleanly, so our expected
+  # start counts are higher than our expected termination counts.
+  if [ "upload" = "$direction" ]; then
+    (( ++expected_ctrl ))
+    (( ++min_expected_start ))
+    (( ++max_expected_start ))
+  fi
+
+  local lines="$(grep "exec: lfs-ssh-echo.*git-lfs-transfer .*${reponame}.git $direction" "$log")"
+  local ctrl_count="$(printf '%s' "$lines" | grep -c -- '-oControlMaster=yes')"
+
+  [ "$expected_ctrl" -eq "$ctrl_count" ]
+
+  assert_ssh_transfer_session_counts "$log" 'spawning pure SSH connection' \
+    "$min_expected_start" "$max_expected_start"
+  assert_ssh_transfer_session_counts "$log" 'pure SSH connection successful' \
+    "$min_expected_start" "$max_expected_start"
+  assert_ssh_transfer_session_counts "$log" 'terminating pure SSH connection' \
+    "$min_expected_end" "$max_expected_end"
+}
+
 begin_test "batch transfers with ssh endpoint (git-lfs-transfer)"
 (
   set -e
@@ -184,13 +231,17 @@ begin_test "batch transfers with ssh endpoint (git-lfs-transfer)"
   git add .gitattributes test.dat
   git commit -m "initial commit"
 
+  # On Windows we do not multiplex SSH connections by default, so we
+  # enforce their use in order to match other platforms' connection counts.
+  git config --global lfs.ssh.autoMultiplex true
+
   GIT_TRACE=1 git push origin main >push.log 2>&1
-  grep "lfs-ssh-echo.*git-lfs-transfer .*$reponame.git upload" push.log
+  assert_ssh_transfer_sessions 'push.log' 'upload' 1 8
   assert_remote_object "$reponame" "$(calc_oid "$contents")" "${#contents}"
 
   cd ..
   GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log
-  grep "lfs-ssh-echo.*git-lfs-transfer .*$reponame.git download" trace.log
+  assert_ssh_transfer_sessions 'trace.log' 'download' 1 8
 
   cd "$reponame-2"
   git lfs fsck

--- a/t/t-batch-transfer.sh
+++ b/t/t-batch-transfer.sh
@@ -116,7 +116,7 @@ begin_test "batch transfers succeed with an empty hash algorithm"
   git add .gitattributes good.dat special.dat
   git commit -m "hi"
 
-  git push origin main 2>&1 | tee push.log
+  git push origin main
   assert_server_object "$reponame" "$(calc_oid "$contents")"
 )
 end_test
@@ -159,7 +159,9 @@ begin_test "batch transfers with ssh endpoint (git-lfs-authenticate)"
   git add .gitattributes test.dat
   git commit -m "initial commit"
 
-  git push origin main 2>&1
+  GIT_TRACE=1 git push origin main >push.log 2>&1
+  [ "1" -eq "$(grep -c "exec: lfs-ssh-echo.*git-lfs-authenticate /$reponame upload" push.log)" ]
+  assert_server_object "$reponame" "$(calc_oid "$contents")"
 )
 end_test
 
@@ -182,10 +184,13 @@ begin_test "batch transfers with ssh endpoint (git-lfs-transfer)"
   git add .gitattributes test.dat
   git commit -m "initial commit"
 
-  git push origin main 2>&1
+  GIT_TRACE=1 git push origin main >push.log 2>&1
+  grep "lfs-ssh-echo.*git-lfs-transfer .*$reponame.git upload" push.log
+
   cd ..
   GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log
   grep "lfs-ssh-echo.*git-lfs-transfer .*$reponame.git download" trace.log
+
   cd "$reponame-2"
   git lfs fsck
 )

--- a/t/t-batch-transfer.sh
+++ b/t/t-batch-transfer.sh
@@ -262,8 +262,8 @@ begin_test "batch transfers with ssh endpoint (git-lfs-transfer)"
   assert_remote_object "$reponame" "$(calc_oid "$contents")" "${#contents}"
 
   cd ..
-  GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log
-  assert_ssh_transfer_sessions 'trace.log' 'download' 1 8
+  GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee clone.log
+  assert_ssh_transfer_sessions 'clone.log' 'download' 1 8
 
   cd "$reponame-2"
   git lfs fsck
@@ -304,8 +304,8 @@ begin_test "batch transfers with ssh endpoint and multiple objects (git-lfs-tran
   assert_remote_object "$reponame" "$(calc_oid "$contents3")" "${#contents3}"
 
   cd ..
-  GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log
-  assert_ssh_transfer_sessions 'trace.log' 'download' 3 8
+  GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee clone.log
+  assert_ssh_transfer_sessions 'clone.log' 'download' 3 8
 
   cd "$reponame-2"
   git lfs fsck
@@ -349,8 +349,8 @@ begin_test "batch transfers with ssh endpoint and multiple objects and batches (
   assert_remote_object "$reponame" "$(calc_oid "$contents3")" "${#contents3}"
 
   cd ..
-  GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log
-  assert_ssh_transfer_sessions 'trace.log' 'download' 3 2
+  GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee clone.log
+  assert_ssh_transfer_sessions 'clone.log' 'download' 3 2
 
   cd "$reponame-2"
   git lfs fsck

--- a/t/t-batch-transfer.sh
+++ b/t/t-batch-transfer.sh
@@ -199,10 +199,32 @@ assert_ssh_transfer_sessions() {
     (( ++max_expected_start ))
   fi
 
+  # Versions of Git prior to 2.11.0 invoke Git LFS via the "smudge" filter
+  # rather than the "process" filter, so a separate Git LFS process runs for
+  # each downloaded object and spawns its own control socket SSH connection.
+  if [ "download" = "$direction" ]; then
+    gitversion="$(git version | cut -d" " -f3)"
+    set +e
+    compare_version "$gitversion" '2.11.0'
+    result=$?
+    set -e
+    if [ "$result" -eq "$VERSION_LOWER" ]; then
+      min_expected_start="$num_objs"
+      max_expected_start="$num_objs"
+      min_expected_end="$num_objs"
+      max_expected_end="$num_objs"
+      expected_ctrl="$num_objs"
+    fi
+  fi
+
+  local max_expected_nonctrl=$(( max_expected_start - expected_ctrl ))
+
   local lines="$(grep "exec: lfs-ssh-echo.*git-lfs-transfer .*${reponame}.git $direction" "$log")"
   local ctrl_count="$(printf '%s' "$lines" | grep -c -- '-oControlMaster=yes')"
+  local nonctrl_count="$(printf '%s' "$lines" | grep -c -- '-oControlMaster=no')"
 
   [ "$expected_ctrl" -eq "$ctrl_count" ]
+  [ "$max_expected_nonctrl" -ge "$nonctrl_count" ]
 
   assert_ssh_transfer_session_counts "$log" 'spawning pure SSH connection' \
     "$min_expected_start" "$max_expected_start"
@@ -242,6 +264,93 @@ begin_test "batch transfers with ssh endpoint (git-lfs-transfer)"
   cd ..
   GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log
   assert_ssh_transfer_sessions 'trace.log' 'download' 1 8
+
+  cd "$reponame-2"
+  git lfs fsck
+)
+end_test
+
+begin_test "batch transfers with ssh endpoint and multiple objects (git-lfs-transfer)"
+(
+  set -e
+
+  setup_pure_ssh
+
+  reponame="batch-ssh-transfer-multiple"
+  setup_remote_repo "$reponame"
+  clone_repo "$reponame" "$reponame"
+
+  contents1="test1"
+  contents2="test2"
+  contents3="test3"
+  git lfs track "*.dat"
+  printf "%s" "$contents1" >test1.dat
+  printf "%s" "$contents2" >test2.dat
+  printf "%s" "$contents3" >test3.dat
+  git add .gitattributes test*.dat
+  git commit -m "initial commit"
+
+  sshurl=$(ssh_remote "$reponame")
+  git config lfs.url "$sshurl"
+
+  # On Windows we do not multiplex SSH connections by default, so we
+  # enforce their use in order to match other platforms' connection counts.
+  git config --global lfs.ssh.autoMultiplex true
+
+  GIT_TRACE=1 git push origin main >push.log 2>&1
+  assert_ssh_transfer_sessions 'push.log' 'upload' 3 8
+  assert_remote_object "$reponame" "$(calc_oid "$contents1")" "${#contents1}"
+  assert_remote_object "$reponame" "$(calc_oid "$contents2")" "${#contents2}"
+  assert_remote_object "$reponame" "$(calc_oid "$contents3")" "${#contents3}"
+
+  cd ..
+  GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log
+  assert_ssh_transfer_sessions 'trace.log' 'download' 3 8
+
+  cd "$reponame-2"
+  git lfs fsck
+)
+end_test
+
+begin_test "batch transfers with ssh endpoint and multiple objects and batches (git-lfs-transfer)"
+(
+  set -e
+
+  setup_pure_ssh
+
+  reponame="batch-ssh-transfer-multiple-batch"
+  setup_remote_repo "$reponame"
+  clone_repo "$reponame" "$reponame"
+
+  contents1="test1"
+  contents2="test2"
+  contents3="test3"
+  git lfs track "*.dat"
+  printf "%s" "$contents1" >test1.dat
+  printf "%s" "$contents2" >test2.dat
+  printf "%s" "$contents3" >test3.dat
+  git add .gitattributes test*.dat
+  git commit -m "initial commit"
+
+  sshurl=$(ssh_remote "$reponame")
+  git config lfs.url "$sshurl"
+
+  # On Windows we do not multiplex SSH connections by default, so we
+  # enforce their use in order to match other platforms' connection counts.
+  git config --global lfs.ssh.autoMultiplex true
+
+  # Allow no more than two objects to be transferred in each batch.
+  git config --global lfs.concurrentTransfers 2
+
+  GIT_TRACE=1 git push origin main >push.log 2>&1
+  assert_ssh_transfer_sessions 'push.log' 'upload' 3 2
+  assert_remote_object "$reponame" "$(calc_oid "$contents1")" "${#contents1}"
+  assert_remote_object "$reponame" "$(calc_oid "$contents2")" "${#contents2}"
+  assert_remote_object "$reponame" "$(calc_oid "$contents3")" "${#contents3}"
+
+  cd ..
+  GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log
+  assert_ssh_transfer_sessions 'trace.log' 'download' 3 2
 
   cd "$reponame-2"
   git lfs fsck

--- a/t/t-batch-transfer.sh
+++ b/t/t-batch-transfer.sh
@@ -186,6 +186,7 @@ begin_test "batch transfers with ssh endpoint (git-lfs-transfer)"
 
   GIT_TRACE=1 git push origin main >push.log 2>&1
   grep "lfs-ssh-echo.*git-lfs-transfer .*$reponame.git upload" push.log
+  assert_remote_object "$reponame" "$(calc_oid "$contents")" "${#contents}"
 
   cd ..
   GIT_TRACE=1 git clone "$sshurl" "$reponame-2" 2>&1 | tee trace.log

--- a/t/testhelpers.sh
+++ b/t/testhelpers.sh
@@ -176,6 +176,22 @@ assert_server_object() {
   }
 }
 
+# assert_remote_object() confirms that an object file with the given OID and
+# size is stored in the "remote" copy of a repository
+assert_remote_object() {
+  local reponame="$1"
+  local oid="$2"
+  local size="$3"
+  local destination="$(canonical_path "$REMOTEDIR/$reponame.git")"
+
+  pushd "$destination"
+    local cfg="$(git lfs env | grep LocalMediaDir)"
+    local f="${cfg:14}/${oid:0:2}/${oid:2:2}/$oid"
+    actualsize="$(wc -c <"$f" | tr -d '[[:space:]]')"
+    [ "$size" -eq "$actualsize" ]
+  popd
+}
+
 check_server_lock_ssh() {
   local reponame="$1"
   local id="$2"

